multibody sdf parsing: Load meshes from relative paths

sammy-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8855)
<!-- Reviewable:end -->

sammy-tri commented:
+@amcastro-tri for feature review

---

Review status: 0 of 14 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855#-:-LD8xl0W9cP6I28hDOkZ:b-2qhvtz)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you @sammy-tri!! feature review done.

---

Reviewed 15 of 16 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 295 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDC8sDS6qVLatMvT3EL:-LDC8sDS6qVLatMvT3EM:bmugehw) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L295)):*
> ```cpp
>   // in which to search for files referenced within the SDF file.
>   std::string root_dir = ".";
>   size_t found = full_path.find_last_of("/\\");
> ```

BTW nice!

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 325 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDC7knRA_Gi3Wxn5u0g:-LDC7knRA_Gi3Wxn5u0h:b-lq73yw) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L325)):*
> ```cpp
>       for (uint64_t visual_index = 0; visual_index < link.VisualCount();
>            ++visual_index) {
>         sdf::Visual sdf_visual = detail::ResolveVisualUri(
> ```

BTW `const sdf::Visual`

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.h, line 37 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCAuRI81vhCrbGkB-Q:-LDCAuRI81vhCrbGkB-R:b3gomwr) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/scene_graph_parser_detail.h#L37)):*
> ```objc
> /// not present, the new object will be identical to the original.
> /// See parsers::ResolveFilename() for more detail on this operation.
> sdf::Visual ResolveVisualUri(const sdf::Visual& original,
> ```

BTW, Since this method is so `<mesh>` specific, maybe add a TODO saying we'd like to pass an `sdf::Messh&` instead? (which doesn't exist yet but it's in the works).

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 43 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDC97AX6kj7ueQczkaO:-LDC97AY7hJHIS0Al7vc:b-y0svgr) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L43)):*
> ```cpp
> // Helper to return the child element of `element` named `child_name`.
> // Returns nullptr if not present.
> sdf::ElementPtr MaybeGetChildElement(
> ```

This method is a duplicate of the one above. Sorry, we merged that recently. Also notice that [per this comment](https://github.com/RobotLocomotion/drake/pull/8828#issuecomment-390656088), we decided to avoid using `sdf::ElementPtr` whenever possible.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 219 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCAQOPBz4zjLriOrtL:-LDCAQOPBz4zjLriOrtM:bb104h4) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L219)):*
> ```cpp
>     sdf::ElementPtr mesh_element = MaybeGetChildElement(geom_element, "mesh");
>     if (mesh_element) {
>       sdf::ElementPtr uri_element = MaybeGetChildElement(mesh_element, "uri");
> ```

I'd suggest doing `string filename = GetChildElementValueOrThrow<std::string>(mesh_element, "uri")` in order to enforce users to provide a `<uri>` for each mesh they have (that is, it is a required tag once `<mesh>` was found). That method also throws a more or less neat message to the user.

---

*[multibody/multibody_tree/parsing/test/links_with_visuals.sdf, line 21 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCCkll27VGB0FToNLo:-LDCCkll27VGB0FToNLp:b-n5yy3m) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/links_with_visuals.sdf#L21)):*
> ```
> 	  <mesh>
> 	    <uri>link1_visual2_mesh.obj</uri>
> 	  </mesh>
> ```

Indentation and this need Emacs' "untabify" (`M-x untabify` )

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 213 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBlbVExMq7vcoajh3:-LDCBlbVExMq7vcoajh4:b-uorynp) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L213)):*
> ```cpp
>   EXPECT_EQ(link1_visual_geometry_ids.size(), 2);
> 
>   // TODO(sam.creasey) Verify that the path to the mesh for the second
> ```

I believe we don't even need this test here. It would be enough with just testing that independently from the MBP parser. What about adding it [here](https://github.com/RobotLocomotion/drake/blob/master/multibody/multibody_tree/parsing/test/scene_graph_parser_detail_test.cc#L170)?, where the `geometry::Mesh` object is in scope.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 219 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBOZ677uvBxJf4Ksu:-LDCBOZ677uvBxJf4Ksv:bnnm9im) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L219)):*
> ```cpp
>   // It would be possible to modify geometry::DispatchLoadMessage to
>   // take a DrakeLcmInterface and then scrape the filename out of the
>   // resulting lcmt_viewer_load_robot message, but I don't want to.
> ```

cc'ing @seancurtis-tri on this one.

---

*[multibody/parsers/parser_path_utils.h, line 14 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCCIXS3LHPb_XZjMCY:-LDCCIXS3LHPb_XZjMCZ:b-j8qwvw) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/parsers/parser_path_utils.h#L14)):*
> ```objc
> // verifying that the file exists. If @p file_name is a relative path, this
> // method converts it into an absolute path based on the current working
> // directory, and then, if the file exists, returns the full path. In either
> ```

would you be able to elaborate on the distinction between the last two cases. Sorry, not clear to me.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855#-:-LDCD0P36ATeocoTfnRa:buvsv61)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+@SeanCurtis-TRI for platform review per rotation.

---

Review status: 10 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 325 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDC7knRA_Gi3Wxn5u0g:-LDCKHySBFLBL2rdy9F0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L325)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW `const sdf::Visual`
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.h, line 37 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCAuRI81vhCrbGkB-Q:-LDCKMtnA5oNvFHkYCny:btgc1gs) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/scene_graph_parser_detail.h#L37)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, Since this method is so `<mesh>` specific, maybe add a TODO saying we'd like to pass an `sdf::Messh&` instead? (which doesn't exist yet but it's in the works).
</blockquote></details>

It's pretty tied to the API of MakeGeometryInstanceFromSdfVisual in that it's trying to create a new version of the input argument for that function where applicable.  I guess you could have it operate on `sdf::Mesh`, but then the caller would have to extract it, which it doesn't need to do now.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 43 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDC97AX6kj7ueQczkaO:-LDCL8n40uInv6JU37wA:b-6w5yfe) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L43)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This method is a duplicate of the one above. Sorry, we merged that recently. Also notice that [per this comment](https://github.com/RobotLocomotion/drake/pull/8828#issuecomment-390656088), we decided to avoid using `sdf::ElementPtr` whenever possible.
</blockquote></details>

Fixed to use `sdf::Element*`.  This is a non-const overload of the const version above.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 219 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCAQOPBz4zjLriOrtL:-LDCLKo4C2539FSQ9Igm:bioam0w) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L219)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I'd suggest doing `string filename = GetChildElementValueOrThrow<std::string>(mesh_element, "uri")` in order to enforce users to provide a `<uri>` for each mesh they have (that is, it is a required tag once `<mesh>` was found). That method also throws a more or less neat message to the user.
</blockquote></details>

I think I'd still need to extract a pointer to the element in order to call `Set()` on it eventually though?  If the element is missing it'll be caught in `MakeShapeFromSdfGeometry`, though I could still add a check here if we want.

---

*[multibody/multibody_tree/parsing/test/links_with_visuals.sdf, line 21 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCCkll27VGB0FToNLo:-LDCLiPC-y_q-GBEpW7L:b-2sprib) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/links_with_visuals.sdf#L21)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Indentation and this need Emacs' "untabify" (`M-x untabify` )
</blockquote></details>

Oops, tabs removed.  Indentation should look OK now.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 213 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBlbVExMq7vcoajh3:-LDCM1K0A_vui8zb66Ki:bo313wy) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L213)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I believe we don't even need this test here. It would be enough with just testing that independently from the MBP parser. What about adding it [here](https://github.com/RobotLocomotion/drake/blob/master/multibody/multibody_tree/parsing/test/scene_graph_parser_detail_test.cc#L170)?, where the `geometry::Mesh` object is in scope.
</blockquote></details>

The resolution of the object file is currently being done in `AddModelFromSdfFile` a couple of calls up the stack from the eventual call to `MakeSdfGeometryFromString`.  My rationale for this was to avoid passing the search directory + `PackageMap` through `MakeGeometryInstanceFromSdfVisual` and `MakeShapeFromSdfGeometry`, though I'm open to changing that if we decide that making the code more locally testable is worth the extra args being passed down (it might be).

---

*[multibody/parsers/parser_path_utils.h, line 14 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCCIXS3LHPb_XZjMCY:-LDCNxxQ-G3nW_Zg9Pqx:b-8qwzku) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/parsers/parser_path_utils.h#L14)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

would you be able to elaborate on the distinction between the last two cases. Sorry, not clear to me.
</blockquote></details>

Rewrote the comment.  Hopefully I haven't made it worse.  :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855#-:-LDCP6hQ1rb_U8anzENi:batupn4)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 9 of 14 files reviewed at latest revision, 5 unresolved discussions.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 219 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCAQOPBz4zjLriOrtL:-LDCY3at0wa19TAekzhq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L219)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Oh, I see, sorry. Yes, I believe we want to throw `std::runtime_error` if either `<uri>` is missing or `resolved_name` is empty.
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 41 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCVCCQ49EzD0ylc4-J:-LDCWTBG-W06RIUzCXBM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3cd69adcb30362dc21f93b6b4b411eb851798cac/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L41)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW `... the mutable child element...`
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/test/links_with_visuals.sdf, line 21 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCUKni5_5O85RPD5ne:-LDCYF50C4T5e-LrnWJ-:bsk66cm) ([raw file](https://github.com/robotlocomotion/drake/blob/3cd69adcb30362dc21f93b6b4b411eb851798cac/multibody/multibody_tree/parsing/test/links_with_visuals.sdf#L21)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW indentation problems still and weird little red arrows in Reviewable.
</blockquote></details>

Something is definitely messed up in my emacs configuration.  Hopefully this is fixed now.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 213 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBlbVExMq7vcoajh3:-LDCZkZU2MZlKom1q-2m:b-d0dylx) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L213)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

The other reason to do that is that eventually @seancurtis-tri might want to split the SG parsing from the MBP parsing. I believe path resolution should also work in that case as well. Lets see what he thinks about this.
</blockquote></details>

FWIW there's now a somewhat indirect test for this after making `ResolveVisualUri` throw `std::runtime_error` if the file is not found.  That might be good enough?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
First pass done.

My *biggest* concern is that the refactoring of the file path stuff means that it will now outlive the old parsers and RBT. As such, I'm holding it to a higher standard than when it was uniquely part of that ecosystem. So, I've got a litany of (mostly cosmetic) comments in that regard.

---

Reviewed 8 of 16 files at r1, 1 of 4 files at r2, 7 of 7 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.h, line 32 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDChGqNBfGY5FzA3y8C:-LDChGqNBfGY5FzA3y8D:bhtntjs) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/multibody_tree/parsing/scene_graph_parser_detail.h#L32)):*
> ```objc
>     const sdf::Visual& sdf_visual);
> 
> /// Given an sdf::Visual object representing a <visual> element from
> ```

BTW This seems overly specific; it could just as easily be applied to geometries for collisions. And, quite possibly, other URIs? However, if you want to defer this until we're actually *parsing* collision meshes, then I'm open to that. But how about a TODO?

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 215 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCkC0B3yLdoE5u9oPS:-LDCkC0B3yLdoE5u9oPT:bw1uml6) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L215)):*
> ```cpp
>   std::shared_ptr<sdf::Element> visual_element = original.Element()->Clone();
>   sdf::Element* geom_element = MaybeGetChildElement(
>       visual_element.get(), "geometry");
> ```

BTW I this formatting won't survive clang format. Same for `resolved_name =`.  And I prefer the clang-formatted version.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 228 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCkd2g4t5Zs3wEIRvX:-LDCkd2g4t5Zs3wEIRvY:b-cnauyn) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L228)):*
> ```cpp
>         } else {
>           throw std::runtime_error(
>               std::string(__FILE__) + ": " + __func__ +
> ```

FYI This *screams* please give me a nice `DRAKE_THROW_RUNTIME` function. I'd use it.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 213 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBlbVExMq7vcoajh3:-LDCpBIhA8_1LUV8hSTr:btf6mxn) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L213)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

FWIW there's now a somewhat indirect test for this after making `ResolveVisualUri` throw `std::runtime_error` if the file is not found.  That might be good enough?
</blockquote></details>

I'm willing to defer this until we work out the more general relationship between an SDF file, SceneGraph, and arbitrary geometry sources.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 219 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBOZ677uvBxJf4Ksu:-LDCoTutEH7caPHM0X8X:b-xj408k) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L219)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

cc'ing @seancurtis-tri on this one.
</blockquote></details>

Reviewable is being weird on me -- this cc seems to be floating in the middle of nothing interesting.

---

*[multibody/parsers/BUILD.bazel, line 61 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCbCGg4jnlfLyrPBVb:-LDCbCGg4jnlfLyrPBVc:bwtrndz) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/BUILD.bazel#L61)):*
> ```Python
>         "//common",
>         "@spruce",
>         "@tinydir",
> ```

BTW Doesn't seem that `find_files` depends on tinyxml2 and tinydir.

---

*[multibody/parsers/parser_path_utils.h, line 10 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCcAFdBwVsZ8oa0cW3:-LDCcAFdBwVsZ8oa0cW4:b1qoirn) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.h#L10)):*
> ```objc
> namespace parsers {
> 
> /// Obtains the full path of @file_name. If @p file_name is already a
> ```

BTW I'm not in love with the fact that this file has both "full path" and "fully-qualified name" which, as near as I can tell, are the same thing.

---

*[multibody/parsers/parser_path_utils.cc, line 48 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCceH461JZoqNFFWbU:-LDCceH461JZoqNFFWbV:bxux0jm) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L48)):*
> ```cpp
> 
> namespace {
> // Searches for key p package in package_map. If the key exists, this saves the
> ```

BTW missing `@` on `@p`? Probably just drop it since this isn't doxygen.

---

*[multibody/parsers/parser_path_utils.cc, line 66 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCej0eEebpINP0TKwf:-LDCej0eEebpINP0TKwg:byp0b65) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L66)):*
> ```cpp
> // The unit test that most directly covers this method is:
> // drake/multibody/parsers/test/urdf_parser_test/urdf_parser_test.cc.
> string ResolveFilename(const string& filename, const PackageMap& package_map,
> ```

BTW What happens if filename is an absolute path to begin with?

---

*[multibody/parsers/parser_path_utils.cc, line 68 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCeCun0_7v6uPPR31P:-LDCeCun0_7v6uPPR31Q:b-5qk6fj) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L68)):*
> ```cpp
> string ResolveFilename(const string& filename, const PackageMap& package_map,
>                        const string& root_dir) {
>   spruce::path mesh_filename_spruce;
> ```

BTW Related to the removed "Mesh" warning, `mesh_filename_spruce` should probably be `qualified_filename`.

---

*[multibody/parsers/parser_path_utils.cc, line 90 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCdn-e9hv1M_pVTSRh:-LDCdn-e9hv1M_pVTSRi:byu23is) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L90)):*
> ```cpp
>       mesh_filename_spruce = spruce::path(package_path_string);
> 
>       auto split_raw = raw_filename_spruce.split();
> ```

BTW Isn't `split_raw` exactly `split_filename`.

---

*[multibody/parsers/parser_path_utils.cc, line 108 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCeQV0E7H04X7eVCHl:-LDCeQV0E7H04X7eVCHm:b-lek5z0) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L108)):*
> ```cpp
>     // If root_dir is a relative path, convert it to an absolute path.
>     bool dir_is_relative =
>         !(normalized_root_dir.size() >= 1 && normalized_root_dir[0] == '/');
> ```

BTW `GetFullPath` uses a *different* mechanism to determine if a path is absolute (`substr`). It'd be nice if a single mechanism were used in the span of a single file.

---

*[multibody/parsers/parser_path_utils.cc, line 119 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCes7UB5maPQl1qAmE:-LDCes7UB5maPQl1qAmF:b-5xwv7e) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L119)):*
> ```cpp
>   }
>   if (!mesh_filename_spruce.exists()) {
>     drake::log()->warn("File '{}' could not be found.",
> ```

BTW Rather than deleting *all* of the warnings, I'd take these two warnings and merge them into one message:

"The file '{}' resolved to '{}' and could not be found."  Or some such thing.

---

*[multibody/parsers/parser_path_utils.cc, line 62 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCh6OV0BbRjgKzulOb:-LDCh6OV0BbRjgKzulOc:b-r0j186) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/parsers/parser_path_utils.cc#L62)):*
BTW Wouldn't this be:

>  // namespace

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855#-:-LDCaLh46iJtsyRq0kC8:bitypz6)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 10 of 14 files reviewed at latest revision, 16 unresolved discussions.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.h, line 32 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDChGqNBfGY5FzA3y8C:-LDD4jGM0k7BXQFxRMRy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/multibody_tree/parsing/scene_graph_parser_detail.h#L32)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This seems overly specific; it could just as easily be applied to geometries for collisions. And, quite possibly, other URIs? However, if you want to defer this until we're actually *parsing* collision meshes, then I'm open to that. But how about a TODO?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 215 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCkC0B3yLdoE5u9oPS:-LDD56h03p7Ynx268IPR:b-3unish) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L215)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I this formatting won't survive clang format. Same for `resolved_name =`.  And I prefer the clang-formatted version.
</blockquote></details>

I don't, but as long as our tool is still more hyperactive than the style guide requires, fine.

---

*[multibody/parsers/BUILD.bazel, line 61 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCbCGg4jnlfLyrPBVb:-LDD0R3nCQZCjbRP04i-:b-9xaj7l) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/BUILD.bazel#L61)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Doesn't seem that `find_files` depends on tinyxml2 and tinydir.
</blockquote></details>

`package_map.cc` uses both.

---

*[multibody/parsers/parser_path_utils.h, line 10 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCcAFdBwVsZ8oa0cW3:-LDD12kY43S2x96xC-pz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.h#L10)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I'm not in love with the fact that this file has both "full path" and "fully-qualified name" which, as near as I can tell, are the same thing.
</blockquote></details>

Done.

---

*[multibody/parsers/parser_path_utils.cc, line 48 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCceH461JZoqNFFWbU:-LDD18Vt-gLfr1CdqmVc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L48)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW missing `@` on `@p`? Probably just drop it since this isn't doxygen.
</blockquote></details>

Done.

---

*[multibody/parsers/parser_path_utils.cc, line 66 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCej0eEebpINP0TKwf:-LDD1MRSBWfde-oRF090:br3nuk2) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L66)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW What happens if filename is an absolute path to begin with?
</blockquote></details>

It seems to do what the doxygen in the header states and prepends `root_dir` to it.  This would break absolute paths, though I'm not sure why you'd have an absolute path in an SDF/URDF.   Does it seem worth changing?

---

*[multibody/parsers/parser_path_utils.cc, line 68 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCeCun0_7v6uPPR31P:-LDD2Kkf3pQTnVBxfH0T:b-niim19) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L68)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Related to the removed "Mesh" warning, `mesh_filename_spruce` should probably be `qualified_filename`.
</blockquote></details>

I went with `full_filename_spruce` since I dropped the "qualified" language due to the defect raised regarding the doxygen.

---

*[multibody/parsers/parser_path_utils.cc, line 90 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCdn-e9hv1M_pVTSRh:-LDD2lMt7utuVNgYqZjx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L90)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Isn't `split_raw` exactly `split_filename`.
</blockquote></details>

Done.

---

*[multibody/parsers/parser_path_utils.cc, line 108 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCeQV0E7H04X7eVCHl:-LDD3Jo84x1q7krDlhuM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L108)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW `GetFullPath` uses a *different* mechanism to determine if a path is absolute (`substr`). It'd be nice if a single mechanism were used in the span of a single file.
</blockquote></details>

Done.

---

*[multibody/parsers/parser_path_utils.cc, line 119 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCes7UB5maPQl1qAmE:-LDD3KDl1sO1qRw_Agwe:bsqlz3b) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L119)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Rather than deleting *all* of the warnings, I'd take these two warnings and merge them into one message:

"The file '{}' resolved to '{}' and could not be found."  Or some such thing.
</blockquote></details>

All callers to this function were displaying an error message and throwing an exception already, so having a warning here about being ignored by drake was actually a bit of a lie about what was actually a fatal error.  Put back a more general warning.

---

*[multibody/parsers/parser_path_utils.cc, line 62 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCh6OV0BbRjgKzulOb:-LDD1BVE-IHOBJHUp9FR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/parsers/parser_path_utils.cc#L62)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Wouldn't this be:

>  // namespace
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 215 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCkC0B3yLdoE5u9oPS:-LDD8iEM9blZDgU7L2iy:b-s2tsbw) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L215)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I don't, but as long as our tool is still more hyperactive than the style guide requires, fine.
</blockquote></details>

Thanks for being a good sport.

---

*[multibody/parsers/parser_path_utils.cc, line 66 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCej0eEebpINP0TKwf:-LDD7nAx0nDVRX8MFGIN:bulyhoz) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L66)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

It seems to do what the doxygen in the header states and prepends `root_dir` to it.  This would break absolute paths, though I'm not sure why you'd have an absolute path in an SDF/URDF.   Does it seem worth changing?
</blockquote></details>

There are two principles running through my mind:

1. Never doubt that a user will do *anything*.
2. Putting in an absolute path seems like it *should* work. And, according to Sherm's "Principle of Least Astonishment", if a user used an absolute path, they shouldn't be surprised that it blows up (either by documenting that it blows up, or by having it work).

---

*[multibody/parsers/parser_path_utils.cc, line 119 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCes7UB5maPQl1qAmE:-LDD8Uj14_DDGiyy49nk:bmr8aj9) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L119)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

All callers to this function were displaying an error message and throwing an exception already, so having a warning here about being ignored by drake was actually a bit of a lie about what was actually a fatal error.  Put back a more general warning.
</blockquote></details>

Oops. Still have "Mesh".

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/parsing/scene_graph_parser_detail.cc, line 228 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCkd2g4t5Zs3wEIRvX:-LDDNckf9nSyZ2tiX6BY:b-hw8eem) ([raw file](https://github.com/robotlocomotion/drake/blob/06918c804743a5b70177a6110dd115ad992b64fb/multibody/multibody_tree/parsing/scene_graph_parser_detail.cc#L228)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI This *screams* please give me a nice `DRAKE_THROW_RUNTIME` function. I'd use it.
</blockquote></details>

BTW yes please!! I'd use it too!

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 213 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBlbVExMq7vcoajh3:-LDDO37q50zNbiABsPNe:b-1fjd6l) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L213)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'm willing to defer this until we work out the more general relationship between an SDF file, SceneGraph, and arbitrary geometry sources.
</blockquote></details>

Re: cc'ing @SeanCurtis-TRI on this one.

This is the spot where I was cc'ing you.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 219 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCBOZ677uvBxJf4Ksu:-LDDO7al80fAbL_Y2v55:bc0k4fo) ([raw file](https://github.com/robotlocomotion/drake/blob/8cfe5dacb284e83371a3127b64583518ddeb15a8/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L219)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Reviewable is being weird on me -- this cc seems to be floating in the middle of nothing interesting.
</blockquote></details>

I remember that happening to me like a year ago.... ah! could it be you are not in the right revision?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 12 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/parsers/parser_path_utils.cc, line 66 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCej0eEebpINP0TKwf:-LDDPo5n00Mzq1-8gYCP:blay52k) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L66)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

There are two principles running through my mind:

1. Never doubt that a user will do *anything*.
2. Putting in an absolute path seems like it *should* work. And, according to Sherm's "Principle of Least Astonishment", if a user used an absolute path, they shouldn't be surprised that it blows up (either by documenting that it blows up, or by having it work).
</blockquote></details>

Done.  (I don't know why I thought it would be at all complicated)

---

*[multibody/parsers/parser_path_utils.cc, line 119 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8855#-LDCes7UB5maPQl1qAmE:-LDDPsrU3C4VHQjuH1LA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4a8056aee5ca2ebcb2fa554b19438bfdca3c759a/multibody/parsers/parser_path_utils.cc#L119)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Oops. Still have "Mesh".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 7 files at r4, 2 of 4 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855#-:-LDDjjSf5r6VITYD9oGy:bnfp4nl)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
@seancurtis-tri anything else blocking?

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855#-:-LDDkTTh9IUY-Dnu3sR6:bs9poif)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Oops.  Nope.  <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8855#-:-LDDlRqV2h9QnWm3y50f:b-akb91e)*
<!-- Sent from Reviewable.io -->
